messages/examples/abc.rs

```diff
@@ -0,0 +1,284 @@
+                        notes.push(messages::Note {
+                            frequency: Some(frequency),
+                            duration_ms: (base_duration * length / 2.0) as u32,
+                        });
+                        notes.push(messages::Note {
+                            frequency: None,
+                            duration_ms: (base_duration * length / 2.0) as u32,
+                        });
+                    } else if tie.is_some() {
+                        notes.push(messages::Note {
+                            frequency: Some(frequency),
+                            duration_ms: (base_duration * length) as u32,
+                        });
+                    } else {
+                        notes.push(messages::Note {
+                            frequency: Some(frequency),
+                            duration_ms: (base_duration * length * 9.0 / 10.0) as u32,
+                        });
+                        notes.push(messages::Note {
+                            frequency: None,
+                            duration_ms: (base_duration * length / 10.0) as u32,
+                        });
+                    }
+                }
+                _ => println!("symbol: {:?}", symbol),
+            }
+        }
+    }
+    Ok(notes)
+}
+
+/// Figure out the duration in milliseconds of a length-1 note.
+fn get_base_duration(header: &TuneHeader) -> Result<f32, Report> {
+    let length_field = header
+        .info
+        .iter()
+        .find(|info| info.0 == 'L')
+        .ok_or_else(|| eyre!("Header field L missing"))?;
+    let length = parse_fraction(&length_field.1)?;
+    Ok(length * WHOLE_NOTE_DURATION)
+}
+
+fn get_key_signature(header: &TuneHeader) -> Result<i8, Report> {
+    let key_signature_field = header
+        .info
+        .iter()
+        .find(|info| info.0 == 'K')
+        .ok_or_else(|| eyre!("Header field K missing"))?;
+    key_signature(&key_signature_field.1)
+}
+
+fn parse_fraction(s: &str) -> Result<f32, Report> {
+    let (numerator, denominator) = s
+        .split_once('/')
+        .ok_or_else(|| eyre!("Invalid fraction {}", s))?;
+    let numerator: f32 = numerator.parse()?;
+    let denominator: f32 = denominator.parse()?;
+    Ok(numerator / denominator)
+}
+
+fn get_semitone(note: Note, accidental: Option<Accidental>, key_signature: i8) -> i32 {
+    let accidental = accidental.unwrap_or_else(|| match note {
+        Note::B if key_signature >= 7 => Accidental::Sharp,
+        Note::E if key_signature >= 6 => Accidental::Sharp,
+        Note::A if key_signature >= 5 => Accidental::Sharp,
+        Note::D if key_signature >= 4 => Accidental::Sharp,
+        Note::G if key_signature >= 3 => Accidental::Sharp,
+        Note::C if key_signature >= 2 => Accidental::Sharp,
+        Note::F if key_signature >= 1 => Accidental::Sharp,
+        Note::B if key_signature <= -1 => Accidental::Flat,
```

Comment:
could probably add a newtype which derives PartialOrd, if you wanted to avoid passing the primitive u8 around

Reply:
Could do, but I'm not sure it's really worth it for a type which is only used internally in this example for passing between two functions. If this were part of a serious library I would.